File: frontend/src/api/collection.ts
import { request, AutomationSupport } from "./campaigns";
import { FeatureInput, FeatureResponse } from "./features";

export type Source = {
  scraping: AutomationSupport;
  auto_labeling_supported: boolean;
  capture_available: boolean;
  source_id: string;
  bank: string;
  bank_type: string | null;
  country: string;
  product_name: string;
  product_category: string;
  language: string;
  page_type: string;
  url: string;
  is_example: boolean;
  import_status: string;
  campaign_id: number | null;
  error: string | null;
  attempted_at: string | null;
};
export type Catalog = {
  sources: Source[];
  banks: string[];
  categories: string[];
  data_mode: string;
};
export type ImportResult = {
  source_id: string;
  status: "success" | "existing" | "failed" | "manual_required";
  supported: boolean;
  message: string | null;
  campaign_id: number | null;
  error: string | null;
};
export type Proposal = {
  campaign_id: number;
  token: string;
  engine: string;
  is_demo: boolean;
  values: Partial<FeatureInput>;
  warnings: string[];
  reviewed: boolean;
  created_at: string;
};
export const getSources = () => request<Catalog>("/scraping/sources");
export const deleteSource = (id: string) =>
  request<void>(`/scraping/sources/${encodeURIComponent(id)}`, "DELETE");
export type CaptureMode = "auto" | "capture_only" | "scrape_and_label";
export type NewSource = {
  bank: string;
  product_name: string;
  product_category: string;
  language: string;
  url: string;
};
export const addSource = (data: NewSource) =>
  request<NewSource & { source_id: string }>("/scraping/sources", "POST", data);
export const scrapeSources = (
  source_ids: string[],
  mode: CaptureMode = "auto",
  recapture = false,
) =>
  request<{ results: ImportResult[] }>("/scraping/run", "POST", {
    source_ids,
    mode,
    recapture,
  });
export const getSuggestions = (id: string) =>
  request<Proposal | null>(`/campaigns/${id}/suggestions`);
export const reviewSuggestions = (id: string, token: string, data: FeatureInput) => {
  const { average_paragraph_length: _derived, ...values } = data;
  return request<FeatureResponse>(`/campaigns/${id}/suggestions/review`, "POST", {
    token,
    values,
  });
};

export type Capture = {
  id: string;
  created_at: string;
  has_screenshot: boolean;
  page: ScrapedContent;
};
export const getCaptures = (id: number) =>
  request<Capture[]>(`/scraping/campaigns/${id}/captures`);
export const captureArtifactUrl = (id: string, artifact: string) =>
  `${(import.meta.env.VITE_API_URL || "http://localhost:8000").replace(/\/$/, "")}/api/scraping/captures/${id}/${artifact}`;

export type ScrapedContent = {
  title: string;
  headline?: string | null;
  text: string;
  headings?: string[];
  paragraphs?: string[];
  bullets?: string[];
  tables?: string[];
  scraped_at: string;
  is_demo: boolean;
  warnings: string[];
  source: { product_name: string; language: string; url: string };
};
export const getScrapedContent = (id: number) =>
  request<ScrapedContent | null>(`/scraping/campaigns/${id}/content`);

File: backend/app/schemas/collection.py
from datetime import datetime
from typing import Literal
from pydantic import Field, HttpUrl
from app.schemas.campaign import Schema
from app.schemas.features import FeatureInput
from app.schemas.automation import SourceDefinition


class SourceView(SourceDefinition):
    scraping: dict[str, bool | str | None]
    auto_labeling_supported: bool = False
    capture_available: bool = True
    import_status: str = "Ready"
    campaign_id: int | None = None
    error: str | None = None
    attempted_at: datetime | None = None


class CatalogResponse(Schema):
    sources: list[SourceView]
    banks: list[str]
    categories: list[str]
    data_mode: str


class SourceCreate(Schema):
    bank: str = Field(min_length=1, max_length=120)
    product_name: str = Field(min_length=1, max_length=300)
    product_category: str = Field(min_length=1, max_length=40, pattern=r"^[a-z0-9_]+$")
    language: Literal["Dutch", "French", "English", "Other"]
    url: HttpUrl = Field(max_length=2048)


class ScrapeInput(Schema):
    mode: Literal["auto", "capture_only", "scrape_and_label"] = "auto"
    recapture: bool = False
    source_ids: list[str] = Field(min_length=1, max_length=50)


class ImportResult(Schema):
    supported: bool = False
    message: str | None = None
    source_id: str
    status: Literal["success", "existing", "failed", "manual_required"]
    campaign_id: int | None = None
    error: str | None = None


class BatchResponse(Schema):
    results: list[ImportResult]


class ProposalRead(Schema):
    campaign_id: int
    token: str
    engine: str
    is_demo: bool
    values: dict[str, str | int | bool | None]
    warnings: list[str]
    reviewed: bool
    created_at: datetime


class ReviewInput(Schema):
    token: str
    values: FeatureInput
